### PR TITLE
N-Gram LM on GPU + Transducer greedy decoding (RNN-T, TDT)

### PR DESCRIPTION
**Pull request by artbataev**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [ASR]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage

**Step 1: Build LM for ASR Model (tokenizer-dependent) on domain texts**

```shell
python nemo/scripts/asr_language_modeling/ngram_lm/train_kenlm.py \
      nemo_model_file="nvidia/parakeet-rnnt-1.1b" \
      train_paths=["<train_manifest>"] \
      kenlm_bin_path=$KENLM_BIN_PATH \
      kenlm_model_file=parakeet-rnnt-1.1b_lm-o6.arpa \
      ngram_length=6 \
      preserve_arpa=true \
      save_nemo=true
```

**Step 2: Run Decoding with LM**

```shell
python examples/asr/speech_to_text_eval.py \
   pretrained_name="nvidia/parakeet-rnnt-1.1b" \
   dataset_manifest=<dataset_manifest>  \
   batch_size=32 \
   output_filename=decoded.jsonl \
   rnnt_decoding.strategy="greedy_batch" \
   rnnt_decoding.greedy.ngram_lm_model="parakeet-rnnt-1.1b_lm-o6.arpa.nemo" \
   rnnt_decoding.greedy.ngram_lm_alpha=0.5
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [x] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: tests/collections/asr/test_ngram_lm.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (lilithgrigoryan) [inline tests/collections/asr/test_ngram_lm.py:1]: 2025

File: nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py [hunk 2]
```diff
@@ -0,0 +1,993 @@
...
+_UNK_ID = -3  # Unk
+_SPECIAL_SYMBOLS_MAP = {"<s>": _BOS_ID, "</s>": _EOS_ID, "<unk>": _UNK_ID}
+
+
+def _log_10_to_e(score):
+    """Convert logarithm with base 10 to natural"""
+    return score / np.log10(np.e)
+
+
+class KenLMBatchedWrapper:
```
Comment (lilithgrigoryan) [inline nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py:56]: Add `get_final()` for API consistency with NGPULM

File: nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py [hunk 3]
```diff
@@ -0,0 +1,993 @@
...
+        """
+        device = labels.device
+        batch_size, max_length = labels.shape
+        if labels_lengths is None:
+            labels_lengths = torch.full([batch_size], fill_value=max_length, dtype=torch.int32, device=device)
+        scores = torch.zeros(labels.shape, device=device)
+        states = self.get_init_states(batch_size=batch_size, bos=bos)
+        # NB: It is possible to speedup this algorithm with a custom kernel (no need to retrieve all weights/labels)
+        for i in range(max_length):
+            # TODO(vbataev): support differentiable implementation with Triton
```
Comment (lilithgrigoryan) [inline nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py:828]: Remove or implement TODO

File: nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py [hunk 4]
```diff
@@ -0,0 +1,993 @@
...
+
+        Returns:
+            tensor [B] with final weights for each state
+        """
+        if self._final_resolved:
+            return self.final_weights[states]
+        logging.warning("Final weights are not resolved; using slow implementation")
+        return self._get_final_pytorch(states=states)
+
+    def resolve_final(self):
```
Comment (lilithgrigoryan) [inline nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py:963]: Do with need this in external API? Maybe batter to make it private?

**Comment by lilithgrigoryan** [comment]
Thanks! Some minor comments above!

File: nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py (hunk 4 above)
Comment (artbataev) [reply nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py]: Fixed - made it private

File: nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py (hunk 3 above)
Comment (artbataev) [reply nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py]: Removed TODO. Currently, `score_sentences` / `forward` are experimental. If we find it useful, let's optimize in another iteration.

File: tests/collections/asr/test_ngram_lm.py (hunk 1 above)
Comment (artbataev) [reply tests/collections/asr/test_ngram_lm.py]: done)

File: nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py (hunk 2 above)
Comment (artbataev) [reply nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py:56]: I added `get_final` + tests. Also, I added `eos: bool` option to `forward` and `score_sentences` to add `eos` scores after the sequence (+ tests).

File: nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py [hunk 5]
```diff
@@ -0,0 +1,993 @@
...
+    return score / np.log10(np.e)
+
+
+class KenLMBatchedWrapper:
+    """
+    KenLM model wrapper for single element and batched queries (slow) for reference decoding and testing purposes.
+    """
+
+    @kenlm_required
+    def __init__(self, lm_path: Path | str, vocab_size: int, token_offset: int = DEFAULT_TOKEN_OFFSET):
```
Comment (lilithgrigoryan) [inline nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py:62]: We can add a from_file() method, similar to GPU LM, to allow a faster switch between the two. This is optional, not a blocker.

File: nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py (hunk 4 above)
Comment (lilithgrigoryan) [reply nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py]: thanks!

File: tests/collections/asr/test_ngram_lm.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        scores_ref = kenlm_wrapper.score_sentences(sentences, bos=bos).to(device)
+        scores_lm = n_gpu_lm(
+            labels=pad_sequence([torch.LongTensor(sentence) for sentence in sentences], batch_first=True).to(device),
+            labels_lengths=torch.LongTensor([len(sentence) for sentence in sentences]).to(device),
+            bos=bos,
+        )
+        assert torch.allclose(scores_lm, scores_ref), "Batched scores do not match"
+
+    @pytest.mark.unit
+    def test_save_load_nemo(self, tmp_path, test_data_dir):
```
[the remark this answers is not in the document]
Comment (artbataev) [reply tests/collections/asr/test_ngram_lm.py:159]: Different purposes of tests:
`test_save_load_nemo` - load ARPA, save NeMo, load NeMo
`test_save_load_from_file` - just to test loading `.from_file`

File: tests/collections/asr/test_ngram_lm.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            assert (states1 == states2).all()
+            assert torch.allclose(scores1, scores2)
+
+    @pytest.mark.unit
+    @pytest.mark.skipif(not KENLM_AVAILABLE, reason="KenLM is not available")
+    @pytest.mark.parametrize("device", DEVICES)
+    @pytest.mark.parametrize("bos", [True, False])
+    def test_sentences(self, n_gpu_lm, kenlm_wrapper, bos: bool, device: torch.device):
+        n_gpu_lm = n_gpu_lm.to(device)
+        sentences = [
```
[the remark this answers is not in the document]
Comment (artbataev) [reply tests/collections/asr/test_ngram_lm.py:102]: Added an empty sentence to both `test_sentences` and `test_final` tests

File: nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py (hunk 5 above)
Comment (artbataev) [reply nemo/collections/asr/parts/submodules/ngram_lm/ngram_lm_batched.py:62]: done

**Review by lilithgrigoryan — APPROVED** [review]
LGTM! Thanks!

**Review by andrusenkoau — APPROVED** [review]
Looks good, thank you!
